### PR TITLE
Create a prototype for OPENSSL_rdtsc

### PR DESCRIPTION
**Pull request by kroeckx**
@dot-asm: Does this make sense?

I think so arches have it as long, other as long long.

### CODE REVIEW

**Comment by paulidale** [comment]
I raised this a while back: #3125

File: test/modes_internal_test.c [hunk 1]
```diff
@@ -889,7 +890,7 @@ static void benchmark_gcm128(const unsigned char *K, size_t Klen,
 #ifdef OPENSSL_CPUID_OBJ
     GCM128_CONTEXT ctx;
     AES_KEY key;
-    size_t start, gcm_t, ctr_t, OPENSSL_rdtsc();
+    size_t start, gcm_t, ctr_t;
```
Comment (paulidale) [inline test/modes_internal_test.c]: sizeof(size_t) != sizeof(unsigned long long)

File: crypto/armcap.c [hunk 2]
```diff
@@ -23,7 +24,7 @@ void OPENSSL_cpuid_setup(void)
 {
 }
 
-unsigned long OPENSSL_rdtsc(void)
+unsigned long long OPENSSL_rdtsc(void)
```
Comment (paulidale) [inline crypto/armcap.c]: I'm not convinced that making this always return `unsigned long long` is a good idea.  Each CPU will have a native _best_ size.

**Comment by dot-asm** [comment]
Trouble is that there is no common type suitable for all platforms. Well, there is one thing one can tell for sure, that return value is 64 bits in all 64-bit builds. Another thing one can tell is that if one tries to use `long long` in 32-bit builds, then there are platforms where return value will appear gabled. This leaves us with following question, wouldn't `long` be least troublesome under the circumstances? Yes, `long` would be least troublesome. With reservation that on WIN64 long is 32 bits. So one can as well argue in favour of size_t, or one can simply say that 32 bits is enough [for purposes of short benchmarking, short enough for not having two wraparounds(*), around a second.]

One can also flip the question and aim for harmonization of all assembly implementations to return 64-bit value. Trouble is that not all processors can deliver that, in which cases single wraparound occurrences(*) will be broken. But one can as well choose disregard this. In other words no matter what one does there always is a corner case :-(

To summarize, given subroutine's usage pattern and overall context is not really inappropriate to be a bit opportunistic and just settle for `long` (or `size_t`), but **not** `long long`.

(*) one 32-bit wraparound works, because you ignore borrow anyway;

**Comment by paulidale** [comment]
Or we could define an internal type that represents the return value: `ossl_rdtsc_value` or some such.

I'd also consider a name change: rdtsc is very x86 specific.  `openssl_instruction_counter` or `openssl_fast_counter` ?  This function is private so the possibility exists.

We not only use this as a timer but also as a low grade entropy source in the RNG infrastructure.

**Comment by kroeckx** [comment]
I'm currently leading towards just making it return an uint32_t.

**Comment by dot-asm** [comment]
Another alternative is to simply omit all those #ifdef segments. They are really debugging artefacts, and they are meant to produce meaningful output only in specific and controlled conditions. I won't miss them (since I kept original copies which are more suitable for debugging purposes) and users are better off using openssl speed...

**Comment by kroeckx** [comment]
So I've pushed a new commit changing it to uint32_t.  I just don't have an idea if the assembler versions still work properly.

**Comment by dot-asm** [comment]
As already said, assembler is good up to size_t.

**Comment by dot-asm** [comment]
> As already said, assembler is good up to size_t.

This came out a little bit wrong. *All* assembly modules will work with return type declared up to size_t. [And some 32-bit ones would work with 64-bit return, some will fail...]

**Comment by paulidale** [comment]
It seems like `size_t` might be the way to go.

**Comment by dot-asm** [comment]
Either way works [including removal of #ifdefs]. I'm giving it few day hours in case Kurt changes his mind, and then approving it.
